src/parser/fragments.rs

```diff
@@ -6,33 +6,2101 @@ use nom::number::complete::{le_f32, le_i16, le_i32, le_i8, le_u16, le_u32, le_u8
 use nom::sequence::tuple;
 use nom::IResult;
 
-use super::decode_string;
+use super::{decode_string, StringReference};
 
 #[derive(Debug, Clone, Copy)]
-pub struct FragmentRef<T>(pub i32, PhantomData<T>);
+pub enum FragmentRef<T> {
+    Name(StringReference, PhantomData<T>),
+    Index(u32, PhantomData<T>),
+}
+
+impl<T> FragmentRef<T> {
+    pub fn new(idx: i32) -> FragmentRef<T> {
+        match StringReference::new(idx) {
+            Some(name_ref) => FragmentRef::Name(name_ref, PhantomData),
+            None => FragmentRef::Index(idx as u32, PhantomData),
+        }
+    }
+}
```

Comment:
The value of the index determines the type of reference this is. When the index is negative (and therefor a valid `StringReference`) then this is a name-based `FragmentRef`. Otherwise it's an index-based `FragmentRef`.